File: framework/languages/dotnet/e2e/RegistryMessaging/Client/Scenarios/RmC9BackpressureScenario.cs
using RegistryMessaging.Client.Support;
using RegistryMessaging.Shared;
using Zlink.HttpClient;

namespace RegistryMessaging.Client.Scenarios;

// RM-C9 verifies that one-way send pressure does not expose a public send
// completion oracle and that the channel recovers for a later request.
internal static class RmC9BackpressureScenario
{
    private const int SlowSendCount = 8;

    public static async Task RunAsync(ZLinkHttpClient backpressureConsumer, ZLinkHttpClient providerA)
    {
        await backpressureConsumer.Post("/profile/backpressure/reset").SubmitAsync<object>();
        var marker = $"rm-c9-{Guid.NewGuid():N}";
        var outcomes = await Task.WhenAll(Enumerable.Range(0, SlowSendCount)
            .Select(index => SendBackpressureCommandAsync(
                backpressureConsumer,
                $"rm-c9-slow-{marker}-{index}")));
        ScenarioAssert.That(
            outcomes.All(outcome => outcome == "Submitted"),
            "RM-C9 expected all one-way sends to be submitted without a public bounded-failure oracle.");

        await Task.Delay(TimeSpan.FromSeconds(10));
        var followUp = (await backpressureConsumer.Post("/profile/request")
            .Body(new ProfileReq("rm-c9-after"))
            .SubmitAsync<ProfileRes>()).Body;
        ScenarioAssert.That(followUp.Value == "profile:rm-c9-after",
            "RM-C9 follow-up request failed after backlog cleared.");

        var evidence = (await providerA.Post("/evidence/wait")
            .Body(new EvidenceWaitReq("rm-c9-after", 20000))
            .SubmitAsync<string[]>()).Body;
        ScenarioAssert.That(
            evidence.Any(line => line.Contains("rm-c9-after", StringComparison.Ordinal)),
            "RM-C9 recovery evidence missing.");
        Console.WriteLine("scenario RM-C9 passed");
    }

    private static async Task<string> SendBackpressureCommandAsync(ZLinkHttpClient backpressureConsumer,
        string commandId)
    {
        return (await backpressureConsumer.Post("/profile/backpressure/send")
            .Body(new ProfileMsg(commandId))
            .SubmitAsync<string>()).Body;
    }
}

File: framework/languages/node/e2e/RegistryMessaging/Client/Scenarios/rm-c9-backpressure-scenario.ts
// RM-C9: backpressure / HWM 포화 시나리오를 검증한다.
import type { ProfileRes } from '../../Shared/messages';
import { postJson } from '../Support/http-client';
import { ensure, uniqueMarker } from '../Support/scenario-assert';

const slowSendCount = 8;

export async function runRmC9(backpressureConsumerUrl: string, providerAUrl: string): Promise<void> {
  await postJson(backpressureConsumerUrl, '/profile/backpressure/reset');
  const marker = uniqueMarker('rm-c9');
  const outcomes = await Promise.all(
    Array.from({ length: slowSendCount }, (_, index) =>
      postJson<string>(
        backpressureConsumerUrl,
        '/profile/backpressure/send',
        { commandId: `rm-c9-slow-${marker}-${index}` }
      ))
  );
  ensure(
    outcomes.every((outcome) => outcome === 'Submitted'),
    'RM-C9 expected all one-way sends to be submitted without a public bounded-failure oracle.'
  );

  await new Promise((resolve) => setTimeout(resolve, 10000));
  const followUp = await postJson<ProfileRes>(
    backpressureConsumerUrl,
    '/profile/request',
    { value: 'rm-c9-after' }
  );
  ensure(followUp.value === 'profile:rm-c9-after', 'RM-C9 follow-up request failed after backlog cleared.');

  const evidence = await postJson<string[]>(
    providerAUrl,
    '/evidence/wait',
    { contains: 'rm-c9-after', timeoutMilliseconds: 20000 }
  );
  ensure(
    evidence.some((line) => line.includes('rm-c9-after')),
    'RM-C9 recovery evidence missing.'
  );
  console.log('scenario RM-C9 passed');
}
